runtime/src/main/java/org/corfudb/runtime/view/replication/ChainReplicationProtocol.java

```diff
@@ -0,0 +1,155 @@
+package org.corfudb.runtime.view.replication;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.protocols.logprotocol.StreamData;
+import org.corfudb.protocols.logprotocol.StreamedLogData;
+import org.corfudb.protocols.wireprotocol.ILogData;
+import org.corfudb.runtime.exceptions.OverwriteException;
+import org.corfudb.runtime.view.Layout;
+import org.corfudb.util.CFUtils;
+import org.corfudb.util.serializer.Serializers;
+
+import javax.annotation.Nonnull;
+import java.util.Map;
+import java.util.UUID;
+
+/**
+ * Created by mwei on 4/6/17.
+ */
+@Slf4j
+public class ChainReplicationProtocol extends AbstractReplicationProtocol {
+
+    final Layout layout;
+
+    public ChainReplicationProtocol(IHoleFillPolicy holeFillPolicy, Layout layout) {
+        super(holeFillPolicy);
+        this.layout = layout;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void write(long globalAddress,
+                      @Nonnull Map<UUID, StreamData> entryMap) throws OverwriteException {
+        int numUnits = layout.getSegmentLength(globalAddress);
+
+        ByteBuf b = Unpooled.buffer();
+        try {
+            // To reduce the overhead of serialization, we serialize only the
+            // first time we write, saving when we go down the chain.
+            StreamedLogData sld = new StreamedLogData(entryMap);
+            sld.getSerializedForm(Serializers.CORFU::serialize, b);
+
+            for (int i = 0; i < numUnits; i++) {
+                log.trace("Write[{}]: chain {}/{}", layout, i + 1, numUnits);
+                // In chain replication, we write synchronously to every unit
+                // in the chain.
+                try {
+                    CFUtils.getUninterruptibly(
+                            layout.getLogUnitClient(globalAddress, i)
+                                    // TODO: Currently, this call won't work until the log unit client is refactored.
+                                    .write(globalAddress, null, null, b, null)
+                            , OverwriteException.class);
+                } catch (OverwriteException oe) {
+                    // Some other wrote here (usually due to hole fill)
+                    // We need to invoke the recovery protocol, in case
+                    // the write wasn't driven to completion.
+                    recover(globalAddress);
```

Comment:
not critical, but good to have. I have not thought through what happens when we see a lot of these happening at the same time.